Index: src/main/java/at/mjst/finbase/desktop/controller/bind/GenericCellValueFactory.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/main/java/at/mjst/finbase/desktop/controller/bind/GenericCellValueFactory.java	(revision 62ba01e3064458c4efab3e6390d37341136c3b02)
+++ src/main/java/at/mjst/finbase/desktop/controller/bind/GenericCellValueFactory.java	(revision )
@@ -24,14 +24,14 @@
 {
     @NonNls
     private static final String ERR_DATA_TYPE = "Field datatype mismatch for '%s'";
-    private final Class<T> dataType;
+    private final Class<?> dataType;
     private FieldIdentifier identifier;
 
     /**
      * @param identifier field identifier assigned to column
      * @param dataType   expected field data type
      */
-    public GenericCellValueFactory(FieldIdentifier identifier, Class<T> dataType)
+    public GenericCellValueFactory(FieldIdentifier identifier, Class<?> dataType)
     {
         this.identifier = identifier;
         this.dataType = dataType; // ToDo: maybe check for data type not necessary?
